In [1]:
import pandas as pd

In [24]:


search_df = pd.read_excel(r"C:\Users\aakar\OneDrive\Desktop\Data Analytics Bootcamp\Python\search_data.csv.xlsx")
search_df.head()


,search_request_id,search_request_created_at,estimated_distance,estimated_duration
0,f6ee886c-0363-4fee-9dcb-649e496bbcde,2025-01-05 00:00:01,16803,1929
1,32de8972-992e-43f4-a619-3207c4ac2847,2025-01-05 00:00:02,43353,3059
2,b82fb857-e164-4915-af21-2ebf5f27a31b,2025-01-05 00:00:02,20568,2045
3,6aca8150-bc96-46e2-8302-2ed374f1ee5a,2025-01-05 00:00:04,12551,1410
4,723b361b-fdac-40f3-a0f2-426c487e6fde,2025-01-05 00:00:04,3875,539


In [25]:
search_df['search_request_id'].count()


np.int64(1048575)

In [26]:


quote_df = pd.read_excel(r"C:\Users\aakar\OneDrive\Desktop\Data Analytics Bootcamp\Python\quote_data.csv.xlsx")
quote_df.head()


,quote_id,search_request_id,driver_id,quote_created_at,distance_to_pickup,duration_to_pickup,driver_rating,estimated_fare
0,9f1cc555-cb22-4f7c-9528-b891fb60cfaf,073ca341-f68f-45b8-b11b-cbac10e6a993,82e067e3-28e6-41f1-a22c-b10f3ce4120b,2025-01-04 18:30:01,235,22,3,117
1,a0cdc15f-f70c-43e0-8f44-b8018e1d2582,d924c6e3-4566-435a-9f73-1340d9a0fd03,ace0ac98-ed05-4d77-875d-f2718493e8fe,2025-01-04 18:30:01,734,136,4.8,630
2,c1f7987a-e718-4c8b-a4a4-4137102b125d,784ae6fd-73e0-487c-bf17-914aa6ff575c,e7c4d775-46e1-490a-819b-8bf63715f61a,2025-01-04 18:30:02,1144,105,4.6,69
3,884e86a8-54f9-485c-a172-d3e361cc4200,34d24bb0-bd2b-405a-9586-6e8656286361,006c969d-bf10-4473-8fd5-f2906457a63e,2025-01-04 18:30:03,227,26,4.8,209
4,04d83b1b-6d21-47f3-9424-e17513f8c7e9,c32f2804-94f5-4e7b-bb63-62066a5567f5,86dfa47a-16e5-4cb2-a0d9-5d4aef3431c0,2025-01-04 18:30:04,214,42,4.7,209


In [27]:
quote_df['quote_id'].count()


np.int64(1048575)

In [28]:
import pandas as pd

booking_df = pd.read_excel(r"C:\Users\aakar\OneDrive\Desktop\Data Analytics Bootcamp\Python\booking_data.csv.xlsx")
booking_df.head()


,booking_id,status,booking_created_at,rider_id,quote_id
0,3ff1b448-4bf3-44b3-9f29-dc3f1d6b4553,CANCELLED,2025-01-05 00:00:04,488b86c4-ac1f-45f9-9bef-0bee9c233d85,e8a85d66-4a5f-4908-98a8-e71ac9a2a6b9
1,ece1beaf-f92d-40e1-a638-40b2a6890175,COMPLETED,2025-01-05 00:00:10,03c927e1-7295-44bf-ac90-0cf56d2d4d59,c8f78a65-43bf-49d7-a2d3-488b7197ad7c
2,f222c7b6-461e-4ecf-b312-5e368dbb341e,COMPLETED,2025-01-05 00:00:12,cbc892f5-e5ab-4d41-b099-07ade6a8aa48,672011d1-8f31-4d5b-8365-bc887e91718f
3,0895d183-e2f6-4e43-8390-27c5c18d3ec1,COMPLETED,2025-01-05 00:00:13,fcae9fdd-1562-4d76-b666-0461848d4e41,f9bdd549-5884-4f6b-9017-1ea19b351aab
4,4ecb4ae9-8682-4cd2-8d57-ee8cbe6560de,COMPLETED,2025-01-05 00:00:13,e5897a0a-4caf-4972-87f0-d4391a0ec58e,0d422590-9777-4719-bb4d-8af5cfe5ec4a


In [29]:
booking_df['booking_id'].count()


np.int64(1048575)

In [30]:
import pandas as pd

cancelled_df = pd.read_excel(r"C:\Users\aakar\OneDrive\Desktop\Data Analytics Bootcamp\Python\booking_cancellation_data.csv.xlsx")
cancelled_df.head()


,booking_id,source,cancelled_at,reason_code
0,0281e1f5-efdd-4c5b-bbe9-fb1371241907,ByUser,2025-01-09 09:54:47,WAIT_TIME_TOO_LONG
1,0349cc7f-f4e0-41fe-b47c-a0e69462737b,ByDriver,2025-01-09 09:54:47,PICKUP_TOO_FAR
2,08ff82d8-aac2-4e40-aed2-343cd6d9ee32,ByDriver,2025-01-09 09:54:47,CUSTOMER_WAS_RUDE
3,0bd5697f-b7b0-4eb7-b72c-a06edf7a7936,ByDriver,2025-01-09 09:54:47,CUSTOMER_WAS_RUDE
4,0c590e7f-24f6-4f1b-854e-724f1b8b7471,ByDriver,2025-01-09 09:54:47,PICKUP_TOO_FAR


In [31]:
cancelled_df['booking_id'].count()


np.int64(477921)

In [36]:
complete_count= booking_df['booking_id'].count()-cancelled_df['booking_id'].count()
complete_count

np.int64(570654)

1.1 Quote conversion from searches

In [107]:
quote_conversion_rate = (
    quote_df['search_request_id'].nunique() * 100) / search_df['search_request_id'].nunique()

quote_conversion_rate


86.61993658059748

Booking conversion from quote

In [ ]:
booking_conversion_rate = (
    booking_df['quote_id'].nunique() * 100) / quote_df['quote_id'].nunique()

booking_conversion_rate


99.96709820470639

Confirmed conversion from booking

In [40]:
completed_conversion_rate=(complete_count*100)/booking_df['booking_id'].count()
completed_conversion_rate

np.float64(54.42185823617767)

1.2 segmenting with respect to time

In [45]:
import pandas as pd

search_df['search_request_created_at'] = pd.to_datetime(
    search_df['search_request_created_at']
)

hour = search_df['search_request_created_at'].dt.hour

search_df['time_segment'] = 'Night'

search_df.loc[(hour >= 6) & (hour < 10), 'time_segment'] = 'Morning'
search_df.loc[(hour >= 10) & (hour < 17), 'time_segment'] = 'Day'
search_df.loc[(hour >= 17) & (hour < 21), 'time_segment'] = 'Evening'

search_df[['search_request_created_at', 'time_segment']]


,search_request_created_at,time_segment
0,2025-01-05 00:00:01,Night
1,2025-01-05 00:00:02,Night
2,2025-01-05 00:00:02,Night
3,2025-01-05 00:00:04,Night
4,2025-01-05 00:00:04,Night
...,...,...
1048570,2025-01-10 14:39:59,Day
1048571,2025-01-10 14:39:59,Day
1048572,2025-01-10 14:39:59,Day
1048573,2025-01-10 14:39:59,Day


In [46]:
import pandas as pd

quote_df['quote_created_at'] = pd.to_datetime(
    quote_df['quote_created_at']
)

hour = quote_df['quote_created_at'].dt.hour

quote_df['time_segment'] = 'Night'

quote_df.loc[(hour >= 6) & (hour < 10), 'time_segment'] = 'Morning'
quote_df.loc[(hour >= 10) & (hour < 17), 'time_segment'] = 'Day'
quote_df.loc[(hour >= 17) & (hour < 21), 'time_segment'] = 'Evening'

quote_df[['quote_created_at', 'time_segment']]


,quote_created_at,time_segment
0,2025-01-04 18:30:01,Evening
1,2025-01-04 18:30:01,Evening
2,2025-01-04 18:30:02,Evening
3,2025-01-04 18:30:03,Evening
4,2025-01-04 18:30:04,Evening
...,...,...
1048570,2025-01-09 07:17:01,Morning
1048571,2025-01-09 07:17:01,Morning
1048572,2025-01-09 07:17:01,Morning
1048573,2025-01-09 07:17:02,Morning


In [49]:
import pandas as pd

booking_df['booking_created_at'] = pd.to_datetime(
    booking_df['booking_created_at']
)

hour = booking_df['booking_created_at'].dt.hour

booking_df['time_segment'] = 'Night'

booking_df.loc[(hour >= 6) & (hour < 10), 'time_segment'] = 'Morning'
booking_df.loc[(hour >= 10) & (hour < 17), 'time_segment'] = 'Day'
booking_df.loc[(hour >= 17) & (hour < 21), 'time_segment'] = 'Evening'

booking_df[['booking_created_at', 'time_segment']]


,booking_created_at,time_segment
0,2025-01-05 00:00:04,Night
1,2025-01-05 00:00:10,Night
2,2025-01-05 00:00:12,Night
3,2025-01-05 00:00:13,Night
4,2025-01-05 00:00:13,Night
...,...,...
1048570,2025-01-11 12:54:40,Day
1048571,2025-01-11 12:54:40,Day
1048572,2025-01-11 12:54:40,Day
1048573,2025-01-11 12:54:40,Day


In [50]:
import pandas as pd

cancelled_df['cancelled_at'] = pd.to_datetime(
    cancelled_df['cancelled_at']
)

hour = cancelled_df['cancelled_at'].dt.hour

cancelled_df['time_segment'] = 'Night'

cancelled_df.loc[(hour >= 6) & (hour < 10), 'time_segment'] = 'Morning'
cancelled_df.loc[(hour >= 10) & (hour < 17), 'time_segment'] = 'Day'
cancelled_df.loc[(hour >= 17) & (hour < 21), 'time_segment'] = 'Evening'

cancelled_df[['cancelled_at', 'time_segment']]


,cancelled_at,time_segment
0,2025-01-09 09:54:47,Morning
1,2025-01-09 09:54:47,Morning
2,2025-01-09 09:54:47,Morning
3,2025-01-09 09:54:47,Morning
4,2025-01-09 09:54:47,Morning
...,...,...
477916,2025-01-07 06:56:45,Morning
477917,2025-01-07 06:56:45,Morning
477918,2025-01-07 06:56:45,Morning
477919,2025-01-07 06:56:45,Morning


In [ ]:
import pandas as pd

search_df['trip_distance_km'] = search_df['estimated_distance']

search_df['trip_type'] = 'Short'

search_df.loc[
    (search_df['trip_distance_km'] >= 5) & (search_df['trip_distance_km'] <= 15),
    'trip_type'
] = 'Medium'

search_df.loc[
    search_df['trip_distance_km'] > 15,
    'trip_type'
] = 'Long'

search_df[['estimated_distance', 'trip_distance_km', 'trip_type']]


,estimated_distance,trip_distance_km,trip_type
0,16.803,16.803,Long
1,43.353,43.353,Long
2,20.568,20.568,Long
3,12.551,12.551,Medium
4,3.875,3.875,Short
...,...,...,...
1048570,3.352,3.352,Short
1048571,1.229,1.229,Short
1048572,20.728,20.728,Long
1048573,6.984,6.984,Medium


1.3 max drop in segment

In [55]:
import pandas as pd

drops = {
    'Search_to_Quote': 1 - quote_conversion_rate,
    'Quote_to_Booking': 1 - booking_conversion_rate,
    'Booking_to_Completed': 1 - completed_conversion_rate
}

drop_series = pd.Series(drops)

max_drop_segment = drop_series.idxmax()
max_drop_value = drop_series.max()

max_drop_segment, max_drop_value


('Booking_to_Completed', -53.42185823617767)

1.4 proposed hypotheses:

peak vs non peak hours
estimated fare impact
driver eta
pickup distance for cancellation
driver acceptance time
trip length
new vs repeat users

In [97]:
search_counts = (
    search_df
    .groupby("time_segment")["search_request_id"]
    .count()
    .reset_index(name="search_request_count")
    .sort_values("search_request_count", ascending=False)
)

search_counts


,time_segment,search_request_count
0,Day,472091
3,Night,397596
2,Morning,137945
1,Evening,40943


In [110]:
trip_type_counts = (
    search_df
    .groupby("trip_type")["search_request_id"]
    .count()
    .reset_index(name="quote_count")
    .sort_values("quote_count", ascending=False)
    

)

trip_type_counts


,trip_type,quote_count
1,Medium,503145
2,Short,376041
0,Long,169389


Question no 2, 

overall cancellation rate

In [58]:
total_bookings = booking_df['booking_id'].count()
cancelled_bookings = cancelled_df['booking_id'].count()

cancellation_rate = cancelled_bookings / total_bookings
cancellation_rate_pct = (cancelled_bookings / total_bookings) * 100
cancellation_rate_pct

np.float64(45.57814176382233)

breakdown by source of cancellation

In [60]:
cancellations_by_type = (
    cancelled_df['source']
    .value_counts()
)
cancellations_by_type

source
ByDriver         256970
ByUser           220214
ByMerchant          619
ByApplication       118
Name: count, dtype: int64

top 3 reasons to cancel vy driver

In [ ]:
import pandas as pd

driver_cancelled_df = cancelled_df[cancelled_df['source'] == 'ByDriver']

top_3_reasons = (
    driver_cancelled_df['reason_code']
    .value_counts()
    .head(3)
    .index
)
top_reason_df = driver_cancelled_df[
    driver_cancelled_df['source'].isin(top_3_reasons)
]
top_reason_df


,booking_id,source,cancelled_at,reason_code,time_segment


In [ ]:
import pandas as pd

driver_cancelled_df = cancelled_df[cancelled_df['source'] == 'ByDriver']

top_3_reasons = (
    driver_cancelled_df['reason_code']
    .value_counts()
    .head(3)
    .index
)
top_3_reasons


Index(['TRAFFIC_JAM', 'PICKUP_TOO_FAR', 'VEHICLE_ISSUE'], dtype='object', name='reason_code')

In [ ]:
import pandas as pd

cancelled_df['is_driver_cancelled'] = (
    cancelled_df['source'] == 'ByDriver'
)

rating_cancel_df = quote_df.merge(
    cancelled_df[['booking_id', 'is_driver_cancelled']],
    on='booking_id',
    how='left'
)

rating_cancel_df['is_driver_cancelled'] = (
    rating_cancel_df['is_driver_cancelled'].fillna(False)
)

rating_cancel_df['rating_bucket'] = pd.cut(
    rating_cancel_df['driver_rating'],
    bins=[0, 3, 4, 4.5, 5],
    labels=['<3', '3–4', '4–4.5', '4.5–5'],
    include_lowest=True
)

analysis_df = (
    rating_cancel_df
    .groupby('rating_bucket')
    .agg(
        total_bookings=('booking_id', 'count'),
        driver_cancellations=('is_driver_cancelled', 'sum')
    )
    .reset_index()
)

analysis_df['driver_cancellation_rate_pct'] = (
    analysis_df['driver_cancellations'] / analysis_df['total_bookings'] * 100
)

analysis_df


KeyError: 'booking_id'

In [73]:
print(cancelled_df.columns.tolist())



['booking_id', 'source', 'cancelled_at', 'reason_code', 'time_segment', 'is_driver_cancelled']


In [74]:
print(quote_df.columns.tolist())


['quote_id', 'search_request_id', 'driver_id', 'quote_created_at', 'distance_to_pickup', 'duration_to_pickup', 'driver_rating', 'estimated_fare', 'time_segment', 'trip_distance_km', 'trip_type']


In [ ]:
cancelled_df['is_driver_cancelled'] = cancelled_df['source'].eq('ByDriver')

driver_cancellation_rate = cancelled_df['is_driver_cancelled'].mean()

driver_cancellation_rate


np.float64(0.5376830061872151)

Question no 3 ,

most critical funnel stage 

Search → Quote

Quote → Booking

Booking → Trip start

Trip start → Trip complete

Now, combining Q1 & Q2:

Driver cancellations mostly happen after booking is confirmed but before trip start.

Rider cancellations may happen right after quote or booking.

Peak time segments and trip types cause driver cancellations.

Therefore, the stage with the highest drop-off / friction is:

Booking → Trip Start

que 3.4

as we know the top reasons are TRAFFIC_JAM', 'PICKUP_TOO_FAR', 'VEHICLE_ISSUE

so to minimize these we can do better search matches according to geo location if pick up is too far, or we may give some incentives as well to complete the long pickup rides so that drivers earn enough and dont mind the pickup distance ,Penalize repeated far-pickup acceptances that cancel

Heatmap of traffic-heavy zones → proactive driver repositioning,Use real-time traffic routing before driver acceptance,Auto-reassign if ETA spikes after acceptance

Mandatory “Ready to Drive” status before going online,Cool-down period after vehicle issue cancellation,Temporary deactivation after repeated vehicle issues......,,,,Incentives for low cancellation streaks

Analyze weekly time-series trends for conversion and cancellations to detect 
seasonality. 

In [ ]:
import pandas as pd

booking_df["booking_created_at"] = pd.to_datetime(booking_df["booking_created_at"])

booking_df["week"] = booking_df["booking_created_at"].dt.to_period("W").astype(str)

weekly = booking_df.groupby("week").agg(
    total_bookings=("booking_id", "count"),
    completed_count=("status", lambda x: (x == "COMPLETED").sum()),
    cancelled_count=("status", lambda x: (x == "CANCELLED").sum())
).reset_index()

weekly["completed_conversion_rate"] = (
    weekly["completed_count"] * 100 / weekly["total_bookings"]
)

weekly["cancellation_rate"] = (
    weekly["cancelled_count"] * 100 / weekly["total_bookings"]
)

weekly.sort_values("week")


,week,total_bookings,completed_count,cancelled_count,completed_conversion_rate,cancellation_rate
0,2024-12-30/2025-01-05,156003,107911,47747,69.172388,30.606463
1,2025-01-06/2025-01-12,892572,589085,303458,65.998597,33.998154


The completed conversion rate declines week-over-week (from ~69.2% to ~66.0%), while the cancellation rate increases (from ~30.6% to ~34.0%), indicating a deterioration in booking completion quality.

The second week has significantly higher volume, suggesting that increased demand may be stressing supply, leading to more cancellations.

This pattern points to volume-driven operational strain, not a demand issue, reinforcing the need for capacity and traffic-aware matching during high-demand weeks.

question no 4,

FUNNEL AND STORYTELLING

In [ ]:
import numpy as np
import plotly.graph_objects as go


quote_df["rating_segment"] = pd.cut(
    quote_df["driver_rating"],
    bins=[0, 3, 4, 5],
    labels=["Low (≤3)", "Mid (3–4)", "High (4–5)"]
)

stages = ["Quotes", "Bookings", "Completed"]

fig_segmented = go.Figure()

for segment in quote_df["rating_segment"].dropna().unique():
    df_seg = quote_df[quote_df["rating_segment"] == segment]

    q = df_seg["quote_id"].nunique()
    b = df_seg["booking_id"].notna().sum()
    c = (df_seg["booking_status"] == "COMPLETED").sum()

    fig_segmented.add_trace(go.Funnel(
        name=str(segment),
        y=stages,
        x=[q, b, c],
        textinfo="value+percent previous"
    ))

fig_segmented.update_layout(
    title="Booking Funnel Segmented by Driver Rating",
    funnelmode="group"
)

fig_segmented.show()


ModuleNotFoundError: No module named 'plotly'

In [95]:
quote_counts = (
    quote_df
    .groupby(["driver_rating", "driver_id"])["quote_id"]
    .count()
    .reset_index(name="quote_count").sort_values("quote_count", ascending=False)
)

quote_counts


,driver_rating,driver_id,quote_count
56594,4.7,5e8afa56-e603-4213-84c2-1df6c9021fd2,162
78498,4.8,24d6b161-f87f-4a0c-9fb8-bb5d2b73a6a0,162
46615,4.7,0c291347-d865-4a0e-9587-5cc313c2b54c,161
77816,4.8,1ba688fb-1660-4d67-8767-2537b77beecf,160
65244,4.7,a6cd7a2a-adf5-49e2-a9b5-e5a34a7dc7ef,155
...,...,...,...
6,0.0,00e3626f-2a58-49b0-af3f-4fe6066950b7,1
104807,5.0,ff95842c-bdef-4686-80f3-7934bc176a61,1
104808,5.0,ffaf7811-f176-47e2-889c-80f3b4e12b34,1
1,0.0,00786ab7-3158-41ca-a8bd-500a0ccbde86,1


insight about driver_rating and driver_id 

High activity is concentrated among high-rated drivers (4.7–4.8)

Top quote counts (155–162) are almost entirely from drivers rated ≥ 4.7

These drivers are likely:
Online more often,More responsive,Preferred by the matching system


Many drivers (including 5.0-rated) have only 1 quote

Indicates:

New drivers,Inactive drivers,Drivers rarely matched due to location / availability



Drivers with rating 0.0 and 1 quote:
Likely unrated (new drivers),Or incomplete data

detailed summary:

search is done max in day followed by night
max searches got medium followed by short and long rides.
max cancellation done by driver followed by user
cancellation reason code is TRAFFIC_JAM followed by PICKUP_TOO_FAR and VEHICLE_ISSUE
completed conversion rate declines week-over-week (from ~69.2% to ~66.0%)
while the cancellation rate increases (from ~30.6% to ~34.0%), indicating a deterioration in booking completion quality.
The second week has significantly higher volume, suggesting that increased demand may be stressing supply, leading to more cancellations.
And not to forget 
        the stage with the highest drop-off / friction is:
        Booking → Trip Start
        meaning driver cancelled max rides irresepective of reason where booking was done but there was no conversion or trip did not start
this makes it the area where improvement is needed, either by encouraging drivers with the incentives based on different drivers slab or short trip acceptance, or also we can restrict the short distance ride continuously to any particular driver in a day, use geo location based matching for short driver eta, and there should be accurate estimated time shown to the user on their ui/ux,proper driver movement should be shown to customer for the trust building regarding the movement of driver resulting in less cancellation from the user end, penalize drivers in case of failure in reaching within the estimated time 

in the long run , drivers should be prioritized by us to meet the supply of the demand, however users should also be taken care of which result in direct revenue generation of the platform and also this will help in users retention.